<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Preprint: treedata.table • treedata.table</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Preprint: treedata.table">
<meta property="og:description" content="treedata.table">
<meta property="og:image" content="https://ropensci.github.io/treedata.table/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">treedata.table</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/AA_treedata.table_intro_english.html">Getting Started With The treedata.table Package</a>
    </li>
    <li>
      <a href="../articles/AB_treedata.table_intro_spanish.html">Breve Introducción al Paquete treedata.table</a>
    </li>
    <li>
      <a href="../articles/B_multiphylo_treedata.table.html">Working With multiPhylo Objects in treedata.table</a>
    </li>
    <li>
      <a href="../articles/C_PartialMatching.html">Partially Matching of Trait Data and Tree(s) in treedata.table</a>
    </li>
    <li>
      <a href="../articles/D_AdditionalFunctions_treedata.table.html">Additional Functions for Manipulating Data in treedata.table</a>
    </li>
    <li>
      <a href="../articles/E_Preprint.html">Preprint: treedata.table</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/ropensci/treedata.table/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="E_Preprint_files/header-attrs-2.9/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Preprint: treedata.table</h1>
                        <h4 class="author">Josef Uyeda, Cristian Roman-Palacios, April Wright</h4>
            
            <h4 class="date">08/08/2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/ropensci/treedata.table/blob/master/vignettes/E_Preprint.Rmd"><code>vignettes/E_Preprint.Rmd</code></a></small>
      <div class="hidden name"><code>E_Preprint.Rmd</code></div>

    </div>

    
    
<p><strong>Summary</strong></p>
<ol style="list-style-type: decimal">
<li><p>The number of tips in phylogenetic trees has significantly increased over the last decade. This trend reflects recent advances in next-generation sequencing, accessibility of public data repositories, and the increasing use of phylogenies in many fields.</p></li>
<li><p>Despite R being central to the analysis of phylogenetic data, manipulation of phylogenetic comparative datasets remains slow, complex, and poorly reproducible.</p></li>
<li><p>Here we describe the first R package extending the functionality, syntax, and speed of data.table to explicitly deal with phylogenetic comparative datasets.</p></li>
<li><p>treedata.table significantly increases speed and reproducibility during the data manipulation steps involved in the basic phylogenetic comparative workflow in R.</p></li>
</ol>
<p><strong>Key-words:</strong> character matrix, comparative methods, evolution, phylogenetics</p>
<p><strong>Introduction</strong></p>
<p>The number and size of published phylogenetic trees have exponentially increased over the years (Fig. 1; e.g. Smith <em>et al.</em> 2011; FitzJohn <em>et al.</em> 2014; Smith &amp; Brown 2019). Ongoing biodiversity sequencing efforts have triggered the development of phylogenetic computational methods able to deal with datasets involving hundreds of thousands of taxa (McMahon <em>et al.</em> 2015). For instance, the early development of MAFFT (Katoh 2002) significantly decreased CPU times required to perform sequence alignment on molecular datasets with thousands of species. Similarly, RAxML (Stamatakis 2006), PATHd8 (Tamura <em>et al.</em> 2012), and TreePL (Smith &amp; O’Meara 2012) greatly reduced computational times during the inference and absolute dating of phylogenetic trees involving thousands of species. Given the unprecedented pace at which phylogenetic data is accumulating (Piel <em>et al.</em> 2000; Redelings &amp; Holder 2017), updating the current phylogenetic workflow to cope with the increasing size of phylogenetic trees is now more critical than ever. Attention should be paid to the development of faster, computationally efficient, and more user-friendly implementations in R that further increase reproducibility. This language is now central to research utilizing phylogenetic comparative methods, and many essential packages and educational materials are made available using this language (Harmon 2019).</p>
<p><strong>A short description of data.table</strong></p>
<p>treedata.table heavily relies on data.table, an R package that enables high-performance extended functionality for data tables (Dowle and Srinivasan 2020). data.table is not only faster than other packages implemented in R, but also is significantly more efficient than tools in other languages such as Python and Julia (<a href="https://h2oai.github.io/db-benchmark/"><u>https://h2oai.github.io/db-benchmark/</u></a>). In addition to speed, data.table has a syntactic structure that is clear and simple to follow. Only three elements are basic to data.table’s primary function: <em>DT<span class="math display">\[i, j, by\]</span></em>. First, the <em>i</em> section is used to specify the rows to be considered in filtering or subsetting operations. Second, the <em>j</em> section indicates the changes happening in the columns (e.g. adding new ones, changing existing ones). Third, the <em>by</em> section is used to perform operations based on grouping variables. A brief but more exhaustive introduction to data.table can be found in the data.table’s vignette and wiki.</p>
<p><strong>The treedata.table workflow</strong></p>
<p>treedata.table is a wrapper for data.table for phylogenetic analyses that matches a phylogeny to the data.table (Table 1). After an initial tree/data matching step, treedata.table continuously preserves the tree/data matching across data.table operations. treedata.table also allows users to run functions from other phylogenetic packages on the processed treedata.table objects. Below, we briefly explain the general workflow under treedata.table.</p>
<ol style="list-style-type: decimal">
<li><p><strong>Tree and character matrix matching:</strong> Using the treedata.table package begins with creating a treedata.table object. as.treedata.table function matches the tip.labels of the phylogeny to a column of names in the data.frame.</p></li>
<li><p><strong>treedata.table operations:</strong> two main functions allow users to make changes to treedata.table objects. Changes are reciprocal between trees and data.</p></li>
</ol>
<!-- --><ol style="list-style-type: upper-alpha">
<li><p><strong>Explicitly dropping taxa:</strong> Taxa in treedata.table objects can be dropped using the droptreedata.table function. Dropped taxa results are removed from the character matrix and trees.</p></li>
<li><p><strong>Data operations:</strong> The most powerful functionality of treedata.table is related to functions calling data.table. The <code><a href="https://rdrr.io/r/base/Extract.html">[</a></code> function, taking the same arguments as the analog function in data.table, can be used to subset rows, select, and/or compute statistics on columns in the character matrix of the treedata.table object (<em>DT[i, j, by]</em>). Operations changing the number of rows in the character matrix will also affect the corresponding taxa in the tree.</p></li>
</ol>
<ol start="3" style="list-style-type: decimal">
<li><p><strong>Data extraction from treedata.table objects:</strong> Users can independently extract trees and character matrices from treedata.table objects using the pulltreedata.table function. The <code><a href="https://rdrr.io/r/base/Extract.html">$</a></code> operator is also a valid alternative to pulltreedata.table. Two additional functions (<code><a href="https://rdrr.io/r/base/Extract.html">[[</a></code> and <code>extractVectors</code>) can be used to extract named vectors from treedata.table objects. These operations streamline formatting of data into the various different input requirements of R functions from other phylogenetics packages.</p></li>
<li><p><strong>Using external functions in treedata.table objects:</strong> the tdt function enables users to easily run external functions on treedata.table objects directly. Specifically, tdt passes data and tree attributes from a given treedata.table object as arguments to functions implemented in other packages.</p></li>
<li><p><strong>Additional functions</strong>: treedata.tree includes additional functions to detect and filter character matrixes by character types (continuous or discrete; detectCharacterType, detectAllCharacters, and filterMatrix). Other functions can be used to examine (head, tail, print) and describe (summary) objects of class treedata.table. Finally, two additional functions can be used to inspect and force column and row names in character matrices (hasNames, forceNames).</p></li>
</ol>
<p><strong>Using treedata.table </strong></p>
<p>This brief step-by-step tutorial is based on treeplyr’s <em>Anolis</em> example data, including 100 tips and 11 characters:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ropensci.github.io/treedata.table">treedata.table</a></span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">anolis</span><span class="op">)</span></code></pre></div>
<p>To use all the functionalities in treedata.table, we first construct a treedata.table object using the as.treedata.table function, which performs an exact name match between the tip labels of the tree and the column in the dataset with the most matches.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">td</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.treedata.table.html">as.treedata.table</a></span><span class="op">(</span>tree <span class="op">=</span> <span class="va">anolis</span><span class="op">$</span><span class="va">phy</span>, data <span class="op">=</span> <span class="va">anolis</span><span class="op">$</span><span class="va">dat</span><span class="op">)</span></code></pre></div>
<p>The resulting object can be inspected using the summary(), head(), tail(), and print() functions. For instance, we can see a description of the treedata.tree object using the summary() function:</p>
<p>summary(td)</p>
<p>Next, we can perform data manipulation steps on the resulting treedata.table object. For instance, we can extract the SVL column (snout-vent length) using the following line:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">td</span><span class="op">$</span><span class="va">dat</span><span class="op">[</span>,<span class="st">'SVL'</span><span class="op">]</span></code></pre></div>
<p>A named vector of the same trait (SVL) can also be extracted using <code>td[["SVL"]]</code> or <code><a href="../reference/extractVector.html">extractVector(td, 'SVL')</a></code>. However, extractVector further supports extraction of multiple traits. For instance, the following code will extract two named vectors: one for SVL and another for ecomorph.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/extractVector.html">extractVector</a></span><span class="op">(</span><span class="va">td</span>, <span class="st">'SVL'</span>, <span class="st">'ecomorph'</span><span class="op">)</span></code></pre></div>
<p>The real power in treedata.table is in co-indexing the tree and table based on functions from data.table. We can use data.table syntax to subset the treedata.table object and include only the first representative from each ecomorph in the <em>Anolis</em> dataset.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">td</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">.SD</span>, <span class="fl">1</span><span class="op">)</span>, by <span class="op">=</span> <span class="st">"ecomorph"</span><span class="op">]</span></code></pre></div>
<p>We can also subset the <em>Anolis</em> dataset to include a single species per ecomorph and island:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">td</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">.SD</span>, <span class="fl">1</span><span class="op">)</span>, by <span class="op">=</span> <span class="fu">.</span><span class="op">(</span><span class="va">ecomorph</span>, <span class="va">island</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<p>Furthermore, can create a new variable summarizing SVL+hostility for only Cuban anoles.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">td</span><span class="op">[</span><span class="va">island</span> <span class="op">==</span> <span class="st">"Cuba"</span>, <span class="fu">.</span><span class="op">(</span>Index <span class="op">=</span> <span class="va">SVL</span> <span class="op">+</span> <span class="va">hostility</span><span class="op">)</span><span class="op">]</span></code></pre></div>
<p>While the options for data manipulations are infinite, the matching between the tree and data attributes is always constant. Finally, users can pass data and trees in treedata.table objects as arguments to functions in other packages. For instance, below we use the tdt function in treedata.table to fit a continuous model of trait evolution for SVL in geiger:</p>
<p>tdt(td, fitContinuous(phy, extractVector(td, ‘SVL’), model=“BM”))</p>
<p>All the functions explained above can even handle multiple trees. For instance, below we fit the same model of continuous trait evolution on SVL based on a multiPhylo tree for the <em>Anolis</em> dataset:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">trees</span> <span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">anolis</span><span class="op">$</span><span class="va">phy</span>,<span class="va">anolis</span><span class="op">$</span><span class="va">phy</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">trees</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"multiPhylo"</span>

<span class="va">td</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.treedata.table.html">as.treedata.table</a></span><span class="op">(</span>tree<span class="op">=</span><span class="va">trees</span>, data<span class="op">=</span><span class="va">anolis</span><span class="op">$</span><span class="va">dat</span><span class="op">)</span>

<span class="fu"><a href="../reference/tdt.html">tdt</a></span><span class="op">(</span><span class="va">td</span>, <span class="fu">fitContinuous</span><span class="op">(</span><span class="va">phy</span>, <span class="fu"><a href="../reference/extractVector.html">extractVector</a></span><span class="op">(</span><span class="va">td</span>, <span class="st">'SVL'</span><span class="op">)</span>, model<span class="op">=</span><span class="st">"BM"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>The vignette to treedata.table contains further information on the functions outlined above and in Table 1.</p>
<p><strong>Computational performance</strong></p>
<p><strong><em>Alternatives to treedata.table.</em></strong> To our knowledge, only treeplyr (Uyeda &amp; Harmon 2020) performs simultaneous operations on combined tree/data objects in R. treeplyr is largely based on dplyr (Wickham <em>et al.</em> 2019)<em>.</em> However, keeping trees and data objects separated in the R environment is a more standard practice. Changes to trees and data are typically performed independently using a combination of functions implemented in ape (Paradis &amp; Schliep 2019), base (R Development Core Team 2013), data.table (Dowle &amp; Srinivasan 2019), or in the tidyverse (Wickham et al. 2019).</p>
<p><strong><em>Methods.</em></strong> We used the microbenchmark (Mersman 2019) function under default parameters to compare the performance of functions in treedata.table to other packages. First, we compared the performance in the initial tree/data matching step between treedata.table and treeplyr (treedata.table::as.treedata.table() and treeplyr::make.treedata()). We simulated trees with 10, 40, and 100 tips using rtree function in ape (Paradis &amp; Schliep 2019). Additionally, we generated random character matrices (50 discrete and 50 continuous traits) matching 90% of tips in the tree. Second, we compared the performance of data operations in treedata.table relative to data.table (Dowle &amp; Srinivasan 2019), base (R Development Core Team 2013), treeplyr (Uyeda 2020), and dplyr (Wickham <em>et al.</em> 2019). This time, we simulated trees with 1000, 10000, and 500000 tips using the rtree function in ape. Again, we generated random character matrices (50 discrete and 50 continuous traits) matching 90% of tips in simulated trees. We compared the performance of treedata.table::[, data.table::[ treeplyr::%&gt;%, dplyr::%&gt;%, and the equivalent functions in base when (1) subsampling the full dataset for rows matching a single level in one discrete character, and (2) estimating the sum and mean of two continuous traits based on the groups of a second discrete character. In data.table syntax for this process would be:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">td</span><span class="op">$</span><span class="va">dat</span><span class="op">[</span><span class="va">Disc1</span> <span class="op">==</span> <span class="st">"A"</span>, <span class="fu">.</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">Cont2</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">Cont3</span><span class="op">)</span><span class="op">)</span>, by <span class="op">=</span> <span class="va">Disc10</span><span class="op">]</span></code></pre></div>
<p><strong><em>Results.</em></strong> treedata.table was &gt;400% faster than treeplyr during the initial data/tree matching step (Fig. 2). For instance, combining a dataset with 10 tips to a character matrix of 40 traits (10% of unmatched tips), as.treedata.table takes an average of 12.314 ns (range=8.100–27.479 ns) relative to the 64.198 ns that were needed in treeplyr (range=48.407–166.328 ns). Differences in the performance between these two functions also scale with the number of taxa.</p>
<p>Next, we examined the performance of data operations in treedata.table relative to data.table, base, treeplyr, and dplyr (Fig. 3). We found that the simultaneous processing of phylogenetic trees in treedata.table's compromised the speed of our package by 90% relative to relative to data.table. However, data manipulation in treedata.table (which simultaneously processes phylogenies) is still significantly faster than in other commonly used packages for data manipulation only such as base (&gt;35%), treeplyr (&gt;60%), and dplyr (&gt;90%). The higher performance of treedata.table relative to other functions also increases with the size of the dataset.</p>
<p><strong>Current limitations of treedata.table</strong></p>
<p>The current release of treedata.table can handle phylo and multiPhylo objects. A single character matrix is shared across all the trees in the treedata.table object. Additionally, all the trees and the only character matrix in the same treedata.table object are forced to have the same tip-level sampling. We acknowledge that partial tree/data matching is desirable in some situations. For instance, users may be interested in performing analyses on trees that, despite having different tip-level sampling, partially overlap with a common character matrix. Similarly, users may be interested in using multiple character matrices instead of only one. Future releases of the treedata.table package will focus on relaxing some restrictions on the tree/data matching.</p>
<p><strong>Conclusions</strong></p>
<p>Here we describe the first R package that extends the functionality, syntax, and speed of data.table for performing operations in phylogenetic comparative datasets. treedata.table is expected to increase code reproducibility while simplifying the complexity of scripts. Finally, data manipulation in treedata.table, which is significantly faster than in other commonly used packages, will allow researchers to fast perform data manipulation on large datasets without requiring outstanding computational resources.</p>
<p><strong>Acknowledgments</strong></p>
<p>This package was partially developed during the “Nantucket phylogeny developeR workshop”, organized by Liam J. Revell (NSF DBI-1759940). AMW was supported by an Institutional Development Award (IDeA) from the National Institute of General Medical Sciences of the National Institutes of Health under grant number P2O GM103424-18.The authors thank Hugo Gruson, Kari Norman, and Julia Gustavsen for helpful comments during review in rOpenSci. Heidi E. Steiner revised an early version of the manuscript and assisted with logo design.</p>
<p><strong>Authors’ contributions</strong></p>
<p>A.M.W., J.C.U. conceived the project. C.R.P., A.M.W., and J.C.U. developed the methods. C.R.P. analyzed the data and led the writing. All co-authors assisted with edits and approved publication.</p>
<p><strong>Data accessibility</strong></p>
<p>The treedata.table package and documentation are hosted at <a href="https://github.com/ropensci/treedata.table/"><u>https://github.com/ropensci/treedata.table</u></a>.</p>
<p><strong>References</strong></p>
<blockquote>
<p>Dowle, M. &amp; Srinivasan, A. (2019). data.table: Extension of `data.frame`. R package version 1.12.6. <a href="https://cran.r-project.org/package=data.table"><u>https://CRAN.R-project.org/package=data.table</u></a> (accessed 26 August 2020).</p>
<p>Fitzjohn, R.G., Pennell, M.W., Zanne, A.E., Stevens, P.F., Tank, D.C. &amp; Cornwell, W.K. (2014). How much of the world is woody? <em>Journal of Ecology</em>, <strong>102</strong>, 1266–1272.</p>
<p>Harmon, L.J. 2019. Phylogenetic Comparative Methods. No commercial publisher.</p>
<p>Katoh, K. (2002). MAFFT: a novel method for rapid multiple sequence alignment based on fast Fourier transform. <em>Nucleic Acids Research</em>, <strong>30</strong>, 3059–3066.</p>
<p>Mcmahon, M.M., Deepak, A., Fernández-Baca, D., Boss, D. &amp; Sanderson, M.J. (2015). STBase: One Million Species Trees for Comparative Biology. <em>Plos One</em>, <strong>10</strong>.</p>
<p>Mersmann, O. (2019). microbenchmark: Accurate Timing Functions. R package version 1.4-7. <a href="https://cran.r-project.org/package=microbenchmark"><u>https://CRAN.R-project.org/package=microbenchmark</u></a> (accessed 26 August 2020).</p>
</blockquote>
<blockquote>
<p>Michonneau F., Brown J.W., &amp; Winter D.J. (2016). rotl: an R package &gt; to interact with the Open Tree of Life data. <em>Methods in Ecology and Evolution</em>, <strong>7</strong>, 1476–1481.</p>
</blockquote>
<blockquote>
<p>Paradis, E. &amp; Schliep, K. (2018). ape 5.0: an environment for modern phylogenetics and evolutionary analyses in R. <em>Bioinformatics</em>, <strong>35</strong>, 526–528.</p>
<p>Piel, W.H., Donoghue, M.J., Sanderson, M.J., &amp; Netherlands, L. (2000). TreeBASE: a database of phylogenetic information. In Proceedings of the 2nd International Workshop of Species 2000.</p>
<p>R Core Team (2013). R: A language and environment for statistical computing. R Foundation for Statistical Computing, Vienna, Austria. <a href="https://www.r-project.org/"><u>https://www.R-project.org/</u></a> (accessed 26 August 2020).</p>
<p>Redelings, B.D. &amp; Holder, M.T. (2017). A supertree pipeline for summarizing phylogenetic and taxonomic information for millions of species. <em>PeerJ</em>, <strong>5</strong>.</p>
<p>Smith, S.A. &amp; Brown, J.W. (2018). Constructing a broadly inclusive seed plant phylogeny. <em>American Journal of Botany</em>, <strong>105</strong>, 302–314.</p>
<p>Smith, S.A. &amp; O’Meara, B.C. (2012). treePL: divergence time estimation using penalized likelihood for large phylogenies. <em>Bioinformatics</em>, <strong>28</strong>, 2689–2690.</p>
<p>Smith, S.A., Beaulieu, J.M., Stamatakis, A. &amp; Donoghue, M.J. (2011). Understanding angiosperm diversification using small and large phylogenetic trees. <em>American Journal of Botany</em>, <strong>98</strong>, 404–414.</p>
<p>Stamatakis, A. (2006). RAxML-VI-HPC: maximum likelihood-based phylogenetic analyses with thousands of taxa and mixed models. <em>Bioinformatics</em>, <strong>22</strong>, 2688–2690.</p>
<p>Tamura, K., Battistuzzi, F.U., Billing-Ross, P., Murillo, O., Filipski, A. &amp; Kumar, S. (2012). Estimating divergence times in large molecular phylogenies. <em>Proceedings of the National Academy of Sciences</em>, <strong>109</strong>, 19333–19338.</p>
<p>Uyeda, J.C. &amp; Harmon, L.J. (2020). treeplyr: 'dplyr' Functionality for Matched Tree and Data Objects. R package version 0.1.9. <a href="https://github.com/uyedaj/treeplyr"><u>https://github.com/uyedaj/treeplyr</u></a> (accessed 26 August 2020).</p>
<p>Wickham, H., Averick, M. , Bryan, J., Chang, W., D'Agostino, L., François, M., François, R., Grolemund, G., Hayes, A., Henry, L., Hester, J. Kuhn, M., Pedersen, T.L., Miller, E., Bache, S.M,., Müller, K., Ooms, J., Robinson, D., Seidel, D.P., Spinu, V., Takahashi, K., Vaughan, D., Wilke, C.O., Woo, K., and Yutani H. (2019). Welcome to the tidyverse. Journal of Open Source Software, <strong>4:</strong>1686.</p>
<p>Wickham, H., François, R., Henry, L. &amp; Müller, K. (2019). dplyr: A Grammar of Data Manipulation. R package version 0.8.3. h<a href="https://cran.r-project.org/package=dplyr"><u>ttps://CRAN.R-project.org/package=dplyr</u></a> (accessed 26 August 2020).</p>
</blockquote>
<p><strong>Figures</strong></p>
<p><img src="media/Image1.png" alt="Image 1" style="height: 4in; width:6in;"></p>
<p><strong>Fig. 1.</strong> Temporal change in phylogenetic tree sizes between 1978 and 2020 based on 927 publications for different animal and plant groups. We used a LOESS smoothing to depict the temporal trend in tree size over time. Data was retrieved from the Open Tree of Life using the rotl R package (Michonneau et al. 2016). A linear regression that accounted for lineage identity indicated the significant increase in tree size over time (R<sup>2</sup>=0.2077, <em>p&lt;</em>0.001).</p>
<p><img src="media/Image2.png" alt="Image 2" style="height: 3in; width:6in;"></p>
<p><strong>Fig. 2.</strong> Results for the treedata.table microbenchmark during tree/data matching steps. Estimates of the timing during the tree/data matching steps under treedata.table are shown in relation to treeplyr. We show median and lower/upper quartiles times for the performance of each package.</p>
<p><img src="media/Image3.png" alt="Image 3" style="height: 4in; width:6in;"></p>
<p><strong>Fig. 3.</strong> Results for the treedata.table microbenchmark during data manipulation. We compare the performance of treedata.table against data.table, base, treeplyr, and dplyr. We show median and lower/upper quartiles times for the performance of each package.</p>
<p><strong>Tables</strong></p>
<p><strong>Table 1</strong>. Brief descriptions of the treedata.table functions.</p>
<table class="table">
<thead><tr class="header">
<th align="left">Category</th>
<th align="left">Function</th>
<th align="left">Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">treedata.table object creation</td>
<td align="left">as.treedata.table</td>
<td align="left">Initial step of the workflow in treedata.table. Used to match a character matrix (of class data.frame) to a single (of class phylo) or multiple trees (class multiPhylo).</td>
</tr>
<tr class="even">
<td align="left">Drop taxa from treedata.table objects</td>
<td align="left">droptreedata.table</td>
<td align="left">Drops taxa from a treedata.table object</td>
</tr>
<tr class="odd">
<td align="left">Data manipulation</td>
<td align="left">[</td>
<td align="left">Function for performing data.table operations on an object of class 'treedata.table'</td>
</tr>
<tr class="even">
<td align="left">Data extraction</td>
<td align="left">[[</td>
<td align="left">Function for extract a named vector from an object of class 'treedata.table'</td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="left">extractVector</td>
<td align="left">Returning a named vector from a treedata.table object</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">pulltreedata.table</td>
<td align="left">Returns a character matrix or tree(s) from a treedata.table object</td>
</tr>
<tr class="odd">
<td align="left">Run functions from other packages</td>
<td align="left">tdt</td>
<td align="left">Runs a function on a 'treedata.table' object</td>
</tr>
<tr class="even">
<td align="left">Detect character type</td>
<td align="left">detectCharacterType</td>
<td align="left">Function to detect whether a character is continuous or discrete</td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="left">detectAllCharacters</td>
<td align="left">Applies detectCharacterType over an entire character matrix</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">filterMatrix</td>
<td align="left">Filter a matrix, returning either all continuous or all discrete characters</td>
</tr>
<tr class="odd">
<td align="left">Examine treedata.table objects</td>
<td align="left">summary</td>
<td align="left">Summarizing treedata.table objects</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">print</td>
<td align="left">Print method treedata.table objects</td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="left">head, tail</td>
<td align="left">Return the first or last part of an treedata.table object</td>
</tr>
<tr class="even">
<td align="left">Inspect column/row names</td>
<td align="left">hasNames</td>
<td align="left">Row and column name check</td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="left">forceNames</td>
<td align="left">Force names for rows, columns, or both</td>
</tr>
</tbody>
</table>
</div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Josef Uyeda, Cristian Roman-Palacios, April Wright.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
